plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs

```diff
@@ -0,0 +1,119 @@
+import           Ide.Types                      (PluginDescriptor (..),
+                                                 PluginId,
+                                                 defaultPluginDescriptor)
+import qualified Language.LSP.Types             as LSP
+import           Stan.Analysis                  (Analysis (..), runAnalysis)
+import           Stan.Core.Id                   (Id (..))
+import           Stan.Inspection                (Inspection (..))
+import           Stan.Inspection.All            (inspectionsIds, inspectionsMap)
+import           Stan.Observation               (Observation (..))
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId = (defaultPluginDescriptor plId) {pluginRules = rules recorder}
+
+newtype Log = LogShake Shake.Log deriving (Show)
+
+instance Pretty Log where
+  pretty = \case
+    LogShake log -> pretty log
+
+data GetStanDiagnostics = GetStanDiagnostics
+  deriving (Eq, Show, Generic)
+
+instance Hashable GetStanDiagnostics
+
+instance NFData GetStanDiagnostics
+
+type instance RuleResult GetStanDiagnostics = ()
+
+rules :: Recorder (WithPriority Log) -> Rules ()
+rules recorder = do
+  define (cmapWithPrio LogShake recorder) $
+    \GetStanDiagnostics file -> do
+      hie <- fromJust <$> getHieFile file
+      let enabledInspections = HM.fromList [(LSP.fromNormalizedFilePath file, inspectionsIds)]
+      let analysis = runAnalysis Map.empty enabledInspections [] [hie]
+      return (analysisToDiagnostics file analysis, Just ())
+
+  action $ do
+    files <- getFilesOfInterestUntracked
+    void $ uses GetStanDiagnostics $ HM.keys files
+  where
+    analysisToDiagnostics :: NormalizedFilePath -> Analysis -> [FileDiagnostic]
+    analysisToDiagnostics file = map (observationToDianostic file) . toList . analysisObservations
+    observationToDianostic :: NormalizedFilePath -> Observation -> FileDiagnostic
+    observationToDianostic file (Observation {observationSrcSpan, observationInspectionId}) =
+      ( file,
+        ShowDiag,
+        LSP.Diagnostic
+          { _range = realSrcSpanToRange $ observationSrcSpan,
+            _severity = Just LSP.DsHint,
+            _code = Just (LSP.InR $ unId (inspectionId inspection)),
+            _source = Just "stan",
+            _message = message,
+            _relatedInformation = Nothing,
+            _tags = Nothing
+          }
+      )
+      where
+        inspection = fromJust $ HM.lookup observationInspectionId inspectionsMap
+        -- Lookin similar to Stan CLI output
+        message :: T.Text
+        message =
+          T.unlines $
+            [ " ✲ Name:        " <> inspectionName inspection,
+              " ✲ Description: " <> inspectionDescription inspection,
+              "Possible solutions:"
+            ]
+              ++ map ("  - " <>) (inspectionSolution inspection)
+
+getHieFile :: NormalizedFilePath -> Action (Maybe HieFile)
```

Comment:
This seems like something we should provide alongside the other HIE stuff? WDYT @wz1000 ?

Reply:
At least this needs some comments.

Reply:
@michaelpj For me this function signature looks pretty self-explanatory. What else could be said about it?

Reply:
Here are a list of questions that I have about this function after reading it:
- Why isn't this function provided with the other HIE file functions we have? Is it doing something suspicious? Is there a reason we don't want to provide HIE files? Is it working against the direction of travel with hiedb?
- How does it work? Are we sure we're constructing HIE files correctly? Can the information we get from the rules be inconsistent with what's coming from hiedb? 

I'm sure you thought about a bunch of these things when you were writing it, but all that knowledge is going to be lost unless it gets written down! :)

Reply:
This function was written by @pepeiborra , so I do not know the answers https://github.com/haskell/haskell-language-server/issues/258#issuecomment-1126890590

Reply:
Those are valid questions, and @wz1000 is the right person to review rather than me. Also, perhaps move this function to the `ghcide` package and refactor other pieces of similar code around the `Core/Compile.hs` and `Core/Rules.hs` modules.

Reply:
@wz1000 Please, tell how to refactor this function?

Reply:
> Also, perhaps move this function to the ghcide package and refactor other pieces of similar code around the Core/Compile.hs and Core/Rules.hs modules.

Did you have a chance to look into this refactoring?

Reply:
I thought, I was waiting on suggestion where to place the function.

Now I moved `getHieFile` to Development.IDE.Core.Rules.

As for `refactor other pieces of similar code around the Core/Compile.hs and Core/Rules.hs modules` I do not understand what you are talking about. Can you explain what you want to be performed?

Reply:
I believe that we have code to create HieFile in Core.Rules and/or Core.Compile

It would be nice to merge it with getHieFile in order to ensure that it stays working.

Reply:
You mean create separate rule for HieFile?

Reply:
No, just abstracting and reusing the same code everywhere

Reply:
@pepeiborra I do not know which code should be abstracted.

Reply:
Fair enough